File: 011/Problem11.cs
using System;
using System.Linq;
using System.Reflection.Metadata.Ecma335;

namespace EulerProblems
{
    class Program
    {
        static void Main(string[] args)
        {
            var watch = new System.Diagnostics.Stopwatch();

            watch.Start();
            Problem();
            watch.Stop();

            Console.WriteLine($"Function Took: {watch.Elapsed}");
        }

        private static int VertMaxProduct(int[,] grid, int l)
        {
            var max = 0;
            var m = 20;
            var product = 1;
            for (var i = 0; i < l + 1; i++)
            {
                for (var j = 0; j < m; j++)
                {
                    product = 1;
                    for (var k = 1; k < l; k++)
                    {
                        product *= grid[i + k, j];
                    }

                    if (product > max)
                    {
                        max = product;
                    }
                }
            }

            return max;
        }

        private static int HorzMaxProduct(int[,] grid, int l)
        {
            var max = 0;
            var m = 20;
            var product = 1;
            for (var i = 0; i < l + 1; i++)
            {
                for (var j = 0; j < m; j++)
                {
                    product = 1;
                    for (var k = 1; k < l; k++)
                    {
                        product *= grid[j, i + k];
                    }

                    if (product > max)
                    {
                        max = product;
                    }
                }
            }

            return max;
        }

        private static int DiagAProduct(int[,] grid, int l)
        {
            var answer = 0;
            var n = 20;
            var m = 20;
            var product = 1;
            for (var i = 0; i < n - l + 1; i++)
            {
                for (var j = 0; j < m - l + 1; j++)
                {
                    product = 1;
                    for (var k = 0; k < l; k++)
                    {
                        product *= grid[i + k, j + k];
                    }

                    if (product > answer)
                    {
                        answer = product;
                    }
                }
            }
            return answer;
        }

        private static int DiagBProduct(int[,] grid, int l)
        {
            var answer = 0;
            var n = 20;
            var m = 20;
            var product = 1;
            for (var i = l - 1; i < n; i++)
            {
                for (var j = 0; j < m - l + 1; j++)
                {
                    product = 1;
                    for (var k = 0; k < l; k++)
                    {
                        product *= grid[i - k, j + k];
                    }

                    if (product > answer)
                    {
                        answer = product;
                    }
                }
            }

            return answer;
        }
        private static int[,] ReadLines(string path)
        {
            var grid = new int[20, 20];
            var i = 0;
            foreach (var line in System.IO.File.ReadLines(path))
            {
                var j = 0;
                string[] lineSplit = line.Split(" ");
                int[] lineInt = new int[20];
                foreach (var numStr in lineSplit)
                {
                    grid[i, j] = Int32.Parse(numStr);
                    j++;
                }

                i++;
            }

            return grid;
        }
        private static void Problem()
        {
            var grid = ReadLines(@"C:\Users\woutb\RiderProjects\EulerProblems\EulerProblems\grid.txt");
            var max = VertMaxProduct(grid, 4);
            var tmp = HorzMaxProduct(grid, 4);
            if (tmp > max)
            {
                max = tmp;
            }

            tmp = DiagAProduct(grid, 4);
            if (tmp > max)
            {
                max = tmp;
            }
            tmp = DiagBProduct(grid, 4);
            if (tmp > max)
            {
                max = tmp;
            }
            Console.WriteLine(tmp);
        }
    }
}

File: 011/Problem11.dart
import 'dart:io';

void main() {
  var numLoops = 1000;
  final stopwatch = Stopwatch()..start();
  for (var i = 0; i < numLoops; i++) {
    problem();
  }
  final elapsedTime = stopwatch.elapsed;
  print(elapsedTime);
}

List<String> readLines() {
  // Gets the lines of the grid.txt and returns them as a string list.
  var arr = new List.filled(20, "", growable: false);
  final file = File('grid.txt');
  List<String> lines = file.readAsLinesSync();
  lines.asMap().forEach((key, value) { arr[key] = value; });
  return arr;
}

List<int> lineToIntArr(line) {
  // Returns a single line of file as int array.
  var strArr = line.split(" ");
  var arr = new List.filled(20, 1, growable: false);
  strArr.asMap().forEach((key, value) {
    arr[key] = int.parse(strArr[key].toString());
  });
  return arr;
}

int vertMaxProduct(List<List<int>> grid, int L) {
  var max = 0;
  const m = 20;
  for (var i = 0; i < L + 1; i++) {
    for (var j = 0; j < m; j++) {
      var product = 1;
      for (var k = 1; k < L; k++) {
        product *= grid[i + k][j];
      }
      if (product > max) {
        max = product;
      }
    }
  }
  return max;
}

int hozMaxProduct(List<List<int>> grid, int L) {
  var max = 0;
  const n = 20;
  const m = 20;
  var product = 1;
  for (var i = 0; i < n; i++) {
    for (var j = 0; j < m - L + 1; j++) {
      product = 1;
      var subset = grid[i].sublist(j, j + L);
      for (var k = 1; k < subset.length; k++) {
        product *= subset[k];
      }
      if (product > max) {
        max = product;
      }
    }
  }
  return max;
}

int diagAProduct(List<List<int>> grid, int L) {
  var answer = 0;
  const n = 20;
  const m = 20;
  var product = 1;
  for (var i = 0; i < n - L + 1; i++) {
    for (var j = 0; j < m - L + 1; j++) {
      product = 1;
      for (var k = 0; k < L; k++) {
        product *= grid[i + k][j + k];
      }
      if (product > answer) {
        answer = product;
      }
    }
  }
  return answer;
}

int diagBProduct(List<List<int>> grid, int L) {
  var answer = 0;
  const n = 20;
  const m = 20;
  var product = 1;
  for (var i = L - 1; i < n; i++) {
    for (var j = 0; j < m - L + 1; j++) {
      product = 1;
      for (var k = 0; k < L; k++) {
        product *= grid[i - k][j + k];
      }
      if (product > answer) {
        answer = product;
      }
    }
  }
  return answer;
}

void problem() {
  const gridSize = 20;
  const L = 4;
  var grid = List.generate(gridSize, (index) => List.filled(gridSize, 0, growable: false), growable: false);
  var lines = readLines();

  for (var i = 0; i < 20; i++) {
    grid[i] = lineToIntArr(lines[i]);
  }

  var max = hozMaxProduct(grid, L);
  var tmp = vertMaxProduct(grid, L);
  if (tmp > max) {
    max = tmp;
  }
  tmp = diagAProduct(grid, L);
  if (tmp > max) {
    max = tmp;
  }
  tmp = diagBProduct(grid, L);
  if (tmp > max) {
    max = tmp;
  }
  print(max);
}
